Fix: readWorkbookDetail 관련 오류 (exception 처리, 기삭제된 문제집)

adorableco opened the pull request:
## **PR**

### ✨ 작업 내용

- **readWorkbookDetail 무효 id exception 처리**


1️⃣ 문제집 id가 유효하지 않을 때 (uuid 형식이 아닐 때)
<img width="300" alt="스크린샷 2024-07-10 오후 5 44 34" src="https://github.com/swm-standard/Phote_BE/assets/90397541/a702982a-d079-4f2e-8692-a3d4bb5532b1">

2️⃣ 이미 삭제된 문제집일 때
<img width="300" alt="스크린샷 2024-07-10 오후 5 45 50" src="https://github.com/swm-standard/Phote_BE/assets/90397541/02f44ab2-be9a-4d66-b8d3-8ea32f51af40">

3️⃣ 문제집 id가 path variable에 없을 때
<img width="300" alt="스크린샷 2024-07-10 오후 5 46 26" src="https://github.com/swm-standard/Phote_BE/assets/90397541/48a78594-923b-47c5-8ff0-8b14e035347b">


- **이미 삭제된 문제집 상세 조회 막기**
➡️ 이 부분에서 비즈니스 로직을 Workbook 엔티티 클래스 내부로 넣어봤습니당 (`isdeleted()`)


> ✅ ✅ `SQLRestriction` 을 사용한게 좋은 것 같아서 차용해보려다가 custom으로 만들어둔 `AlreadyDeletedException` 을 사용해야 해서 그냥 deletedAt 체크하는걸로 구현했습니다. (쿼리 요청 횟수는 같더라구요!)

- **삭제 여부 test 완료**
---

### ✨ 참고 사항

#### [c06c9b8](https://github.com/swm-standard/Phote_BE/pull/47/commits/c06c9b8a6c1d47ebf9794ccee17303bbe54192c6) 커밋 관련 참고 사항
(@RinRinPARK question 쪽에서 구현할 때 한번 확인해주세용)

- 문제집 내에서 문제 삭제 여부 확인 후 `QuestionSetDto`로 변환함

> ✅question이 삭제될 경우 해당 questionSet은 함께 soft delete된다고 생각해서 question 실제 삭제 여부는 따지지 않음

- Workbook 내의 QuestionSet 을 dto로 감싸지 않으면 `isDeleted` 메서드가 json에 결과와 함께 출력됨 -> 메서드 제외용 dto를 만듦

> ✅api에 직접적인 연관이 없어서 단순 엔티티명 + dto 로 `QuestionSetDto` 라고 이름 붙임 

---

### ⏰ 현재 버그

x

---

### ✏ Git Close #40

adorableco committed:
Fix: readWorkbookDetail 무효 id exception 처리

part of SWM-60
related to #40

adorableco committed:
Fix: 이미 삭제된 문제집 상세 조회 막기

part of SWM-60
related to #40

adorableco committed:
Refactor: Question 비즈니스 로직을 엔티티 내부로 이동

part of SWM-60
related to #40

adorableco committed:
Test: 문제집 삭제 여부

part of SWM-60
related to #40

adorableco committed:
!HOTFIX: Workbook의 modifiedAt updatable로 변경

part of SWM-60
related to #40

RinRinPARK reviewed (changes requested):
문제집 내 문제 삭제가 된 경우(문제집-문제 관계가 끊어진 경우)에도 readWorkbookDetail 했을 때 문제가 불러와 지네용 
근데 저도 deleteQuestion할 때 QuestionSet 엔티티와의 관계를 끊어주는걸 까먹어서 고치러 가야겠습니다.... ㅋㅋㅋㅋㅋㅋ

adorableco commented:
> 문제집 내 문제 삭제가 된 경우(문제집-문제 관계가 끊어진 경우)에도 readWorkbookDetail 했을 때 문제가 불러와 지네용 근데 저도 deleteQuestion할 때 QuestionSet 엔티티와의 관계를 끊어주는걸 까먹어서 고치러 가야겠습니다.... ㅋㅋㅋㅋㅋㅋ

헉드.. 생각을 못하고 있었군 고칠게욥!!

adorableco committed:
Fix: readWorkbookList 에서 삭제된 문제는 제외

- 문제집 내에서 문제 삭제 여부 확인 후 `QuestionSetDto`로 변환함

✅question이 삭제될 경우 해당 questionSet은 함께 soft delete된다고 생각해서 question 실제 삭제 여부는 따지지 않음

- Workbook 내의 QuestionSet 을 dto로 감싸지 않으면 `isDeleted` 메서드가 json에 결과와 함께 출력됨 -> 메서드 제외용 dto를 만듦
✅api에 직접적인 연관이 없어서 단순 엔티티명 + dto 로 `QuestionSetDto` 라고 이름 붙임

part of SWM-60
related to #40

RinRinPARK reviewed (approved):
확인했습니다 머지해도 될 것 같네용👍
근데 테스트 하다보니(이 pr과는 관계없지만..) deleteWorkbook api호출하면 workbook만 삭제되고 Question Set 엔티티에 정의되어 있던 관계는 삭제되지 않아서 해당 api 수정이 필요할 것 같습니다

adorableco committed:
Fix: QuestionSet deletedAt 필드 var로 변경

✅question이 삭제될 경우 해당 questionSet은 함께 soft delete된다고 생각해서 question 실제 삭제 여부는 따지지 않음

part of SWM-60
related to #40

adorableco commented:
> 확인했습니다 머지해도 될 것 같네용👍 근데 테스트 하다보니(이 pr과는 관계없지만..) deleteWorkbook api호출하면 workbook만 삭제되고 Question Set 엔티티에 정의되어 있던 관계는 삭제되지 않아서 해당 api 수정이 필요할 것 같습니다

맞악 나도 이거 하다가 발견...
